### PR TITLE
[WIP] detect invalid email query conditions

### PR DESCRIPTION
**Pull request by binary1230**
This is part of the way towards fixing https://github.com/magfest/ubersystem/issues/2984

This will cause our system to trip and bail on the condition that's allowing double emails to send.

This does NOT fix the underlying issue, just exposes it and works around it.  There needs to be a part 2 which does something different with the Season supporter emails (which, are actually old and inactive, we could just cut them.  though if we leave the dead code in there, we should fix it so it can't be accidentally re-enabled to cause issues)

### CODE REVIEW

**Comment by binary1230** [comment]
TLDR: this enforces the following condition:

AutomatedEmails.queries[TYPE] must contain ONLY 'TYPE' classes.

So AutomatedEmails.queries[Attendee] must contain only Attendee classes.  Nothing else can contain Attendee classes.

The fact that this wasn't true before was causing the following issue because the pairings looked like this:

```
for <class 'mits.models.MITSTeam'>:
--MITSTeam seen 32 times
for <class 'mivs.models.IndieJudge'>:
--IndieJudge seen 56 times
for <class 'panels.models.PanelApplication'>:
--PanelApplication seen 326 times
for <class 'magprime.models.SeasonPassTicket'>:
--Attendee seen 170 times
for <class 'uber.models.attendee.Attendee'>:
--Attendee seen 16163 times
for <class 'uber.models.group.Group'>:
--Group seen 1163 times
for <class 'hotel.models.Room'>:
--Room seen 5 times
for <class 'mivs.models.IndieStudio'>:
--IndieStudio seen 109 times
for <class 'guests.models.GuestGroup'>:
--GuestGroup seen 82 times
for <class 'mivs.models.IndieGame'>:
--IndieGame seen 107 times
```

Notice the odd man out:
```
for <class 'magprime.models.SeasonPassTicket'>:
--Attendee seen 170 times
```

SeasonPassTicket (which is old) would cause 170 Attendees (who are season supporters > $200 pay) to go through an extra round of being processed by each Email Category.  That means it would go through something like the 'Prereg Payment' category and ask it if it wanted to send that email.

Our caching system, which caches which emails were sent out previously to which attendees, does NOT update DURING the course of one email daemon run (5 minutes).  It assumes that only one unique combo of model instance + email category (i.e. Attendee #42 + "you paid for your reg" email category) will be checked per daemon run.  If it happens twice by accident somehow, normally we would not resend an email that was already sent.  But since the list of cached things that were sent isn't updated during the course of the run for performance reasons, we can currently send emails twice.

I recommend we merge this PR and https://github.com/magfest/magprime/pull/237 as the solution to this issue.  I could use another set of eyes on it before going.  But it should work.

**Comment by RobRuana** [comment]
Okay, after looking over this pull request, along with [this other pull request](https://github.com/magfest/magprime/pull/237), I am getting less and less comfortable with this direction. It feels like we are slapping one band-aid over another, ad infinitum. I'll try to do a quick outline some of my thoughts.


# PREVIOUSLY_SENT_EMAILS
Why don't we keep `c.PREVIOUSLY_SENT_EMAILS` up-to-date every time we send an email?
```python
c.PREVIOUSLY_SENT_EMAILS.add(email.model, email.fk_id, email.ident)
```
This will, of course, be erased whenever we reset the request cache. But that's okay, because the new models will already be saved to the database, so they will be repopulated in the cache the next time the query is run.


# SeasonPassTicket

The `SeasonPassTicket` code seems ill-conceived. I'm having a hard time understanding how it works, or how it was _supposed_ to work. Is this a feature we are currently using? If not, I'd be in favor of ripping it out entirely and replacing it with something more robust down the line.

I don't think we have to worry about it either way right now, as you'll see in my next point. With any luck, we can easily solve the root of the problem, while simultaneously slashing the amount of processing we're doing.


# _assert_same_model_type()

This is an interesting fix. It took a good deal of investigation and cleverness to figure out that there was a disconnect between the instance type returned by our queries, and the class type that we're using to index the queries.

However, I don't know if thats what we _want_ to do. It's a little disconcerting that the attendees returned for the `SeasonPassTicket` were being compared against ALL the other email categories. Moreover, every single model instance returned by every single query is compared against every single  email category. This constitutes exponential growth as the size of our database and the number of emails we have increases.

Consider the following methods from the `` class:
```python
```

**Comment by binary1230** [comment]
yea, cool that approach seems much better than what I was doing.  I sorta knew my brain was fuzzing out last night so, I like your idea better.  can you run with it from here?

**Comment by RobRuana** [comment]
I don't think I'll have time. I already spent too much time on this, and I'm running behind schedule on the autographs work.

**Comment by binary1230** [comment]
For right now we could just disable the season pass stuff in the other PR
and fix the rest after Magfest.  That would stop the dupe emails

**Comment by binary1230** [comment]
hey, if this is solved better in #3068, I'm cool closing this. up to you rob

**Comment by RobRuana** [comment]
Closing as "overcome by events", see #3068.
